Support JSON5 format for parsing .babelrc configuration file

chandlerprall opened the pull request:
According to [Babel documentation](http://babeljs.io/docs/en/babelrc), the `.babelrc` file is serialized as JSON5 which, among other features, allows comments to be included in the file. This PR includes the popular `json5` npm module and uses it to parse the babel configuration file instead of node's built-in `JSON.parse`.

ahnpnl commented:
hi @chandlerprall , not everyone uses `babel` so I don't agree to put `json5` into `dependencies`. This will make the users who don't use `babel` install extra package which they don't need.

I'd prefer to see a warning or error to users ask them to install `json5` so `babel` can use it to parse configuration file.

cc @GeeWee @huafu

huafu commented:
@chandlerprall thanks for the PR! I totally agree with @ahnpnl.
You should add it as a peer dependency, and add a helper in [src/utils/imports.ts](https://github.com/kulshekhar/ts-jest/blob/master/src/utils/imports.ts) to import it the same way.

Also if you could create a new test (duplicating the `babel-config-merge-with-babelrc` for example) that'd be awesome.

GeeWee commented:
Not sure about the peerDependency, that's going to give everyone warnings if they don't have it installed. Unfortunately there's probably not a better solution, as there doesn't seem to be an option for optionalPeerDependency or something like that.

huafu commented:
@GeeWee that is why there is the helpers in the `imports.ts`, we provide understandable error messages for the user to know what they should install in case a peer dep is missing ;-)

ahnpnl commented:
We can do the similar thing like jest did with toMatchInlineSnapshot. They throw the error that the users need to have prettier install in their project and stop the test. I think we can do similar thing.

huafu commented:
@ahnpnl see the comment above https://github.com/kulshekhar/ts-jest/pull/665/comment#issuecomment-411796418

chandlerprall commented:
Thanks for all the feedback! Changes made:

- moved `json5` dependency into _peerDependencies_ and _devDependencies_
- created import util and type for `json5`
- modified `get_babel_rc.ts` to use import util, only if the module is required at runtime (e.g. babel is used)

> Also if you could create a new test (duplicating the babel-config-merge-with-babelrc for example) that'd be awesome.

@huafu I'd previously modified the `use-babelrc` test to include a comment inside its `.babelrc`, testing this functionality. Is there any other specific tests you'd like to see included?

huafu commented on src/types.ts:
```diff
@@ -4,11 +4,13 @@ import * as _babel from '@babel/core';
 export type TBabelPluginIstanbul = typeof _babelIstanbul;
 export type TBabelPresetEnv = typeof _babelEnv;
 export type TBabelPresetJest = typeof _babelJest;
+export type TJson5 = typeof _json5;
```
It has typings so install dev dep. `@types/json5` and use it ;-)

huafu commented on src/utils/get-babel-rc.ts:
```diff
@@ -6,11 +6,21 @@ import {
   {
     basename: BABELRC_FILENAME,
-    read: f => JSON.parse(readFileSync(f, 'utf8')),
+    read: f => parseBabelConfig(readFileSync(f, 'utf8')),
```
I'd actually do a helper maybe in `src/utils/parse-json.ts`, even tho for now we'll use this only for babelrc:

```js
import { importJson5 } from './imports'

let parse: any

export default (data: string) => {
    if (!parse) {
	    try {
	    	return JSON.parse(res);
        } catch(err) {}
        parse = importJson5().parse;
    }
    return parse(data)
}
```

Then change the error message in parseJson to be more generic.

huafu replied:
... so that it loads it ONLY if the default `JSON.parse` would fail

chandlerprall replied:
I disagree with a parse function that does both; I think it's much better to be explicit which JSON spec is intended to be used.

For example, if this function is used for some other file which must be pure JSON (e.g. parsed by another module that enforces JSON-only), it would be better to fail here as well.

If you still want this change I'll add it, but wanted to give my thoughts first.

huafu replied:
I see your point, I guess here it's just a matter of name. It could be named `parseJson5` or so. Then we either use `JSON.parse()` when it has to be pure JSON, else parseJson5.

The main thing here is that users who have a babelrc with valid JSON shouldn't need to install `json5` as a dep...

@ahnpnl @GeeWee thoughts?

ahnpnl replied:
I like the idea of @chandlerprall  to have separate parse functions and depending on which type of JSON the user passes in then the code calls the according parse function to deal with the file.

huafu commented:
@chandlerprall about the test, by having 2 separate tests, the `JSON.parse()` not failing branch would also have been tested, but I guess it's fine especially since `JSON.parse` should not fail haha ;-)

ahnpnl commented:
I still like the way how jest doesn't include optional dependencies into their peerDependencies. Do we have a way to do similar thing ? At the moment we show warning/error message but we still have to add to peerDependencies. There must be a way to do similar like jest does.

huafu commented:
@ahnpnl the peers are in ts-jest, and they are not installed automatically, the user has to install it himself. It's just to tell the user which version we are compatible with, if they are going to install one.

...or did I miss-understood your message?

ahnpnl commented:
No I mean can we do the similar thing like jest requires prettier to use toMatchInlineSnapshot.

So my wondering can we not add peerDependencies but searching for json5 in node_modules and throw errors when cannot find it and stop the whole execution. 

When using toMatchInlineSnapshot from jest, if the project doesn't have prettier installed, jest will stop executing and throw the error that the user has to install it to be able to use that feature.

huafu commented:
@ahnpnl an example for babel: https://github.com/kulshekhar/ts-jest/blob/master/src/utils/imports.ts#L8-L17

We could replace `throw` with `process.exit()`, so that it exit on first attempt to transform a file using ts-jest. Tho what about multi-projects jest setup and such...

Also not to mention that there is not `load` hook with jest, it's on the first file matching the key regex in `transforms` option that the transformer is called and initialized (with NO option/config) if not yet initialized.

ahnpnl commented:
hmm too bad, I wish they had sort of pre transform hook so we can hook it in.

huafu commented:
@ahnpnl I wished that so much too, and actually looked into it for quite some time haha.

chandlerprall commented:
@huafu renaming the function to `parseJson5` solves my concern. I'll make the try...catch change if others agree (or rather, if there are no disagreements :) )

huafu commented:
@chandlerprall ok thanks. Also when I told you to use the typings from `@types/json5`, you should add type def to the result of the `importJson5` using the types provided by it ;-)

chandlerprall commented:
> Also when I told you to use the typings from @types/json5, you should add type def to the result of the importJson5 using the types provided by it ;-)

I tried that and couldn't make TS happy, so I left it doing inference. The [json5 type](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/json5/index.d.ts#L43) is simply

```
declare var json5: JSON;
export = json5;
```

and I'm unsure how to import the `json5` type to reference as the method's return value type. One alternative that works is defining it as `JSON`, which fulfills the contract but doesn't feel correct to me.

huafu commented:
I think keeping the related code as it was before installing `@types/json5`, and replacing:
```ts
import _json5 from 'json5';
```
with
```ts
import * as _json5 from 'json5';
```

should od the trick then

huafu commented:
I was just thinking... it's ugly and it wouldn't be secure if we don't know where the babelrc come from, but parsing a dirty json file can be done like that without any dependency:

```ts
const parseDirtyJson<T>(input: string): T => {
  let r: T;
  eval(`r = ${input};`);
  return r;
}
```

chandlerprall commented:
Pushed the updated typedef, exporting the `typeof _json5` worked great.

chandlerprall commented:
@huafu I've changed to your idea; `parseJson5` utility method that tries `JSON.parse` before falling back to the `json5` module.

huafu commented:
Oh, thanks @chandlerprall. I am very busy lately sorry I haven't looked before. Sounds good to me!

SimenB commented:
Note that Jest includes `babel-core` out of the box (used for mock hoisting and code coverage), meaning that everybody that uses Jest has Babel installed already. So I wouldn't worry about dependency bloat as long as you're in range of whatever Babel uses to read the config

huafu commented:
@SimenB thanks! Totally forgot about that!

`jest` => `jest-cli` => `jest-config` => `babel-core` => `json5`

Sorry @chandlerprall...

chandlerprall commented:
Ha, no worries. Should I undo the `importJson5` wrapper?

huafu commented:
@chandlerprall if you can, that would be good yes. So `json5` in deps, `@types/json5` in devDeps, and use it straight away to parse babel config without any wrapper.

chandlerprall commented:
@huafu  re-updated to use `json5` directly.
